### DISCUSSION TITLE
How can I use Yup with fastify?

### DISCUSSION THREAD

**Discussion by oof2win2** [fastify/fastify#3197] [Q&A] [answered]
So the docs say that to use Yup with fastify, I need to do something like [this](https://www.fastify.io/docs/latest/Validation-and-Serialization/#validator-compiler) (Using other validation libraries). I think that I do have that though
```ts
fastify.get<{
	Params: {
		id: string
	}
}>("/:id", {
	schema: {
		querystring: Yup.object({
			id: Yup.string().required()
		}).required()
	},
	validatorCompiler: ({ schema }) => {
		return (data) => {
			try {
				const result = schema.validateSync(data, yupOptions)
				return { value: result }
			} catch (e) {
				return { error: e }
			}
		}
	}
}, async (req, res) => {
	const { id } = req.params

	const rule = await RuleModel.findOne({ id: id })
	return res.status(200).send(rule)
})
```
This is from my `./routes/rules.ts` file, which I require and add into fastify with `app.ts`. The issue is that `validateSync` does not work, as `Property 'validateSync' does not exist on type 'FastifySchema'.ts(2339)`. The docs don't however say that I need to have something else. What should I do?

**Comment by Eomm** [comment]
What yup version are you using?
The code seems right, but adding a [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) would help a lot to help you.

**Reply by oof2win2** [reply]
Using `yup@0.32.9`.
Here is code that breaks when validating. When the validation middleware is removed, it works fine.
```ts
import Fastify, { FastifyInstance, RouteShorthandOptions } from 'fastify'
import * as Yup from "yup"

const fastify: FastifyInstance = Fastify({})

const yupOptions = {
	strict: false,
	abortEarly: false, // return all errors
	stripUnknown: true, // remove additional properties
	recursive: true
}

fastify.get<{
	Params: {
		id: string
	}
}>("/:id", {
	schema: {
		params: Yup.object({
			id: Yup.string().required()
		}).required()
	},
	validatorCompiler: ({ schema }) => {
		return (data) => {
			try {
				const result = schema.validateSync(data, yupOptions) //this line doesnt work
				return { value: result }
			} catch (e) {
				return { error: e }
			}
		}
	}
}, async (req, res) => {
	const { id } = req.params
	res.send(`You requested with the ID of ${id}`)
})

const start = async () => {
	console.log("a")
	try {
		await fastify.listen(3000)
		console.log(`Server listening on :3000`)

	} catch (err) {
		fastify.log.error(err)
		process.exit(1)
	}
}
start()
```

**Reply by Eomm** [reply]
How do you call the endpoint?

I see that you are validating `querystring` but the URL has the id itself.

**Reply by oof2win2** [reply]
`http://localhost:3000/hereisastring` will respond with `You requested with the ID of hereisastring`
Edit: oh yeah, I am doing that. That however doesn't change anything, it still doesn't work even if I replace `querystring` with `params`. Updated code to reflect what should actually be validated.

Also, in the `validatorCompiler` function, `schema` is of the type [FastifySchema](https://github.com/fastify/fastify/blob/main/types/schema.d.ts#L7), so there is no function `validateSync` on it. I therefore think that the type is the issue but I'm not sure what I can do about it.

**Reply by Eomm** [reply]
I have tried plain JS and it works as expected. The `schema` object is the `Yup.object(...)` in the code

**Reply by oof2win2** [reply]
Hmm, that's weird. As you said, plain JS works. I'll try to figure something out

**Comment by oof2win2** [comment, marked as answer]
Re-declaring `schema` as `any` in the validator seems to work just fine. Not a very nice solution, but it at least works.
```ts
fastify.get<{
	Params: {
		id: string
	}
}>("/:id", {
	schema: {
		params: Yup.object({
			id: Yup.number().required()
		}).required()
	},
	validatorCompiler: ({ schema }: { schema: any }) => {
		return (data) => {
			try {
				const result = schema.validateSync(data, yupOptions) //this line doesnt work
				return { value: result }
			} catch (e) {
				return { error: e }
			}
		}
	}
}, async (req, res) => {
	const { id } = req.params
	res.send(`requested ${id}`)
})
```

**Reply by yahia-soliman** [reply]
Update: you can use `Schema` class from `yup`
```ts
import { Schema, ValidateOptions } from 'yup';

const fastify = Fastify({ logger: true });

const yupOptions: ValidateOptions = {
  strict: false,
  abortEarly: true,
  stripUnknown: true,
  recursive: true,
};
fastify.setValidatorCompiler(({ schema }: { schema: Schema }) => {
  return function (data) {
    try {
      const result = schema.validateSync(data, yupOptions);
      return { value: result };
    } catch (e) {
      return { error: e as Error };
    }
  };
});
```
